src/Native/Unix/System.Security.Cryptography.Native.Apple/pal_x509chain.cpp

```diff
@@ -0,0 +1,262 @@
+{
+    va_list args;
+
+    va_start(args, formatString);
+    CFStringRef fmt = CFStringCreateWithFormatAndArguments(nullptr, nullptr, CFSTR("%@"), args);
+    va_end(args);
+    CFDataRef bleh = CFStringCreateExternalRepresentation(nullptr, fmt, CFStringGetSystemEncoding(), '?');
+
+    if (bleh != nullptr)
+    {
+        printf("%.*s\n", static_cast<int>(CFDataGetLength(bleh)), CFDataGetBytePtr(bleh));
+        CFRelease(bleh);
+    }
+
+    CFRelease(fmt);
+}
+
+extern "C" int32_t
+AppleCryptoNative_X509ChainEvaluate(SecTrustRef chain, double epochDeltaSeconds, bool allowNetwork, int32_t* pOSStatus)
+{
+    if (pOSStatus != nullptr)
+        *pOSStatus = noErr;
+
+    if (chain == nullptr || pOSStatus == nullptr)
+        return -1;
+
+    CFDateRef cfEvaluationTime = CFDateCreate(kCFAllocatorDefault, epochDeltaSeconds);
+    *pOSStatus = SecTrustSetVerifyDate(chain, cfEvaluationTime);
+    CFRelease(cfEvaluationTime);
+
+    if (*pOSStatus != noErr)
+    {
+        return -2;
+    }
+
+    //*pOSStatus = SecTrustSetNetworkFetchAllowed(chain, allowNetwork);
+    (void)allowNetwork;
+
+    if (*pOSStatus != noErr)
+    {
+        return -3;
+    }
+
+    SecTrustResultType trustResult;
+    *pOSStatus = SecTrustEvaluate(chain, &trustResult);
+
+    printf("SecTrustGetTrustResult returned %d and emitted %d\n", *pOSStatus, trustResult);
+
+    if (*pOSStatus != noErr)
+    {
+        return 0;
+    }
+
+    // If: The chain was built with no errors (Unspecified)
+    // Or: The chain was built and involved an explicitly trusted cert (Proceed)
+    // Then: Success.
+    if (trustResult == kSecTrustResultUnspecified || trustResult == kSecTrustResultProceed)
+    {
+        return 1;
+    }
+
+    // Should this be a different return code?
+    return 1;
+}
+
+extern "C" int64_t AppleCryptoNative_X509ChainGetChainSize(SecTrustRef chain)
+{
+    if (chain == nullptr)
+        return -1;
+
+    return SecTrustGetCertificateCount(chain);
+}
+
+extern "C" SecCertificateRef AppleCryptoNative_X509ChainGetCertificateAtIndex(SecTrustRef chain, int64_t index)
+{
+    if (chain == nullptr || index < 0)
+        return nullptr;
+
+    return SecTrustGetCertificateAtIndex(chain, index);
+}
+
+extern "C" CFArrayRef AppleCryptoNative_X509ChainGetTrustResults(SecTrustRef chain)
+{
+    if (chain == nullptr)
+    {
+        return nullptr;
+    }
+
+    CFDictionaryRef detailsAndStuff = SecTrustCopyResult(chain);
+    CFTypeRef detailsPtr = CFDictionaryGetValue(detailsAndStuff, CFSTR("TrustResultDetails"));
+    CFArrayRef details = reinterpret_cast<CFArrayRef>(const_cast<void*>(detailsPtr));
+
+    CFRetain(details);
+    CFRelease(detailsAndStuff);
+
+    printf("Just the details:\n");
+    DumpDict(nullptr, details);
+
+    return details;
+}
+
+static void MergeStatusCodes(CFTypeRef key, CFTypeRef value, void* context)
+{
+    // Windows (and therefore .NET) certificate status codes are defined on an int32_t.
```

Comment:
No... int32_t.  The rest of the comment explains that I'm using a 64-bit number here so I have 32 bits of status values to report to .NET and use the leftovers to report errors from this function in populating the status fields.